### PR TITLE
APEXMALHAR-2063 Made window data manager use file system wal

### PR DESCRIPTION
**Pull request by otterc**
@PramodSSImmaneni @tweise  @ilooner 
Please review

### CODE REVIEW

**Comment by otterc** [comment]
@ilooner Can you please take a look?

**Comment by ilooner** [comment]
@chandnisingh there are conflicts

File: library/src/main/java/org/apache/apex/malhar/lib/wal/WindowDataManager.java [hunk 1]
```diff
@@ -41,15 +41,42 @@
...
+   * Typically it is used to replay tuples of successive windows in input operators after failure.
+   *
+   * @param windowId window id
+   * @return saved state for the window id.
+   * @throws IOException
+   */
+  Object retrieve(long windowId) throws IOException;
+
+  /**
+   * Delete the artifact corresponding to the
```
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/WindowDataManager.java]: complete javadoc here?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/WindowDataManager.java [hunk 2]
```diff
@@ -126,36 +146,24 @@ public void teardown()
...
       return null;
     }
 
     @Override
-    public void delete(int operatorId, long windowId) throws IOException
-    {
-    }
-
-    @Override
-    public void deleteUpTo(int operatorId, long windowId) throws IOException
```
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/WindowDataManager.java:145]: These are no longer needed?

Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/WindowDataManager.java:145]: changed the method to be committed(long windowId). WindowDataManager is Evolving so this was ok.

**Comment by otterc** [comment]
@ilooner I have resolved the conflicts. Please have a look.

**Comment by otterc** [comment]
@ilooner @PramodSSImmaneni @tweise 
This change touches all the operators which use WindowDataManager. Any changes to those operators will cause conflicts again so please review it when you get time.

**Comment by tweise** [comment]
Chandni, yes it is on my TODO list.

## 

sent from mobile

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 3]
```diff
@@ -51,291 +66,607 @@
...
+  private final transient Kryo kryo = new Kryo();
+
+  private transient FileContext fileContext;
 
   public FSWindowDataManager()
   {
-    replayState = TreeMultimap.create();
-    largestRecoveryWindow = Stateless.WINDOW_ID;
-    recoveryPath = DEF_RECOVERY_PATH;
+    kryo.setClassLoader(Thread.currentThread().getContextClassLoader());
```
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:154]: for my learning :) why is this needed?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 4]
```diff
@@ -51,291 +66,607 @@
...
+
+          if (slice == null || !slice.equals(DELETED)) {
+            //delete entry not found so window was not deleted
+            long window = Longs.fromByteArray(windowSlice.toByteArray());
+
+            if (ceilingWindow != null && window > ceilingWindow) {
+              break;
+            }
+            endOffset = offset;
+            lastWindow = window;
```
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: I'm not sure if I'm understanding how delete works correctly. But should the lastWindow be rolled back to the previous window when we encounter DELETED?

Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Would the endoffset have to be rolled back as well?

**Comment by tweise** [comment]
CONFLICT (content): Merge conflict in library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 3 above)
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:154]: Wondering about that also?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 5]
```diff
@@ -51,291 +66,607 @@
...
-    Configuration configuration = new Configuration();
+    operatorId = context.getId();
+
     if (isRecoveryPathRelativeToAppPath) {
-      appPath = new Path(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + recoveryPath);
+      statePath = context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + recoveryPath;
     } else {
-      appPath = new Path(recoveryPath);
+      statePath = recoveryPath;
     }
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:166]: Is statePath shared for all logical operators? What happens if there are 2 operators in a DAG that are using the window data manager?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 6]
```diff
@@ -51,291 +66,607 @@
...
+  public void deleteLastWindow() throws IOException
   {
-    return largestRecoveryWindow;
+    wal.getWriter().append(DELETED);
+    wal.getWriter().flush();
   }
 
   @Override
-  public void partitioned(Collection<WindowDataManager> newManagers, Set<Integer> removedOperatorIds)
+  public void committed(long windowId) throws IOException
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: It would be good to summarize below logic in the javadoc block.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 7]
```diff
@@ -51,291 +66,607 @@
...
+  }
 
-    if (fs.listStatus(new Path(appPath, Integer.toString(operatorId))).length > 0) {
-      long[] windowsAfterReplay = storageAgent.getWindowIds(operatorId);
-      Arrays.sort(windowsAfterReplay);
-      for (long lwindow : windowsAfterReplay) {
-        if (lwindow <= windowId) {
-          storageAgent.delete(operatorId, lwindow);
-        }
+  public Map<Long, Object> retrieveAllWindows() throws IOException
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: No reference to this method? Is it needed. If yes, please add javadoc.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 8]
```diff
@@ -51,291 +66,607 @@
 public class FSWindowDataManager implements WindowDataManager
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:109]: It may be useful to add a few lines WRT dynamic partitioning support and the relationship to checkpointing vs. commit per window.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 9]
```diff
@@ -51,291 +66,607 @@
...
+          //the read only wal can be removed.
+          walIterator.remove();
+        }
+      }
     }
+  }
 
-    return windowIdsArray;
+  @Override
+  public void save(Object object, long windowId) throws IOException
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:409]: Is this method supposed to be called from endWindow only? It appears that it assumes batch operation?

**Comment by tweise** [comment]
@chandnisingh  any chance you can look into it?

**Comment by otterc** [comment]
@tweise will address the review comments by the weekend

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 3 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:154]: I remember that if we don't have this, then Kryo sometimes is not able to find a Class. All the kryo utils that we have `KryoSerializableStreamCodec` and `KryoCloneUtils` do set  class loader on the kryo instance explicitly.

**Comment by otterc** [comment]
Resolved conflicts. Will address pending comments today as well

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 7 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Will move this to `IncrementalCheckpointManager`. `AbstractManagedState` uses it.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 9 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:409]: Yes this is called from  `endWindow()`. FSWindowDataManager does use the wal in batch mode. The wal instance in this class is initialized with `inBatchMode=true`.
The batch mode was added so that multiple entries related to one window go into the same wal file, that is, data of a window is not distributed over multiple wal files.
WindowDataManager provided the support for deleting last saved artifact. Please see `deleteLastWindow()`. `AbstractJMSInputOperator` uses that call to deleted the last artifact in case ack fails.
In the new implementation, when `deleteLastWindow()` is invoked, I just add marker bytes `DELETED` to the file instead of actually truncating the file. 

Therefore, in order to make sure that the wal file doesn't get rotated between save and delete, the batching was added and FSWindowDataManager needs the batch mode to be set to true.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 4 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: @ilooner 
delete is handled by the else block from 283-284.
lastWindow and  endOffset are not at all modified for a window that is marked deleted.

I fail to see the problem. Can you please have another look?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 5 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:166]: Yeah this is a problem. By default `statePath` will be shared by all logical operator. There is another ticket APEXMALHAR-2068 opened for that. Also the fix will need the recent change to Apex Core (APEXCORE-448 ) which will not be there until the core version in malhar changes.

**Comment by otterc** [comment]
Addressed comments

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 9 above)
Comment (tweise) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:409]: @chandnisingh this question isn't directly related to this PR but a potential bug in AbstractJMSInputOperator. If it removes the incomplete state in the finally block and that does not succeed, then what happens on restart? Should this not be done in setup instead?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 10]
```diff
@@ -18,324 +18,718 @@
...
+   *                                      otherwise.
    */
-  public void setRecoveryPathRelativeToAppPath(boolean recoveryPathRelativeToAppPath)
+  public void setStatePathRelativeToAppPath(boolean statePathRelativeToAppPath)
   {
-    isRecoveryPathRelativeToAppPath = recoveryPathRelativeToAppPath;
+    isStatePathRelativeToAppPath = statePathRelativeToAppPath;
   }
+
+  protected static Slice DELETED = new Slice("DEL".getBytes());
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Is this used to indicate that a window segment wasn't complete? If so, then this isn't working because writing of the marker may also fail and we should write the complete marker instead (which is end of window)? Please clarify, maybe I got that wrong?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java [hunk 11]
```diff
@@ -227,6 +240,47 @@ public void setMaxLength(long maxLength)
...
+    return inBatchMode;
+  }
+
+  /**
+   * When in batch mode, a file is rotated only when a batch gets completed. This facilitates writing multiple entries
+   * that will all be written to the same part file.
+   *
+   * @param inBatchMode write in batch mode or not.
+   */
+  public void setInBatchMode(boolean inBatchMode)
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java]: Should this really be exposed as public setter?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowReplayWAL.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+import javax.validation.constraints.NotNull;
+
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.collect.TreeMultimap;
+
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * A FileSystemWAL that WindowDataManager uses to save state of every window.
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowReplayWAL.java]: Minor nit, use link fro friendly javadoc and preserve during refactor:

```
 * A {@link FileSystemWAL} that WindowDataManager uses to save state of every window.
```

File: library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java [hunk 13]
```diff
@@ -293,51 +324,56 @@ public void testTrigger() throws InterruptedException, IOException, TimeoutExcep
...
-    for (int i = 0; i < 12; i++) {
-      File testFile = new File(testMeta.dataDirectory, "file" + i + ".txt");
-      noOfBlocks += (int)Math.ceil(testFile.length() / (2 * 1.0));
-    }
-
-    testMeta.fileSplitterInput.setBlockSize(2L);
-    testMeta.fileSplitterInput.setBlocksThreshold(10);
     testMeta.fileSplitterInput.beginWindow(1);
-
+
```
Comment (tweise) [inline library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java]: trailing whitespace, please check the IDE settings.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java [hunk 14]
```diff
@@ -18,324 +18,718 @@
...
     }
+  }
 
-    return windowIdsArray;
+  @Override
+  public void save(Object object, long windowId) throws IOException
+  {
+    closeReaders();
+    FileSystemWAL.FileSystemWALWriter writer = wal.getWriter();
+    writer.batchCompleted(); //finish previous batch and rotate if needed
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Maybe rotateIfNecessary would be more applicable? Also, it may be good to capture some of this in method javadoc?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 10 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: JMSInput used this to explicitly delete the last saved artifact.
Earlier this used to delete the file  which also could have failed and not have actually deleted the file.  Even though the delete fails, JMSInput will not emit duplicate because if the entry failed to be deleted then in replay the message from file is added to a `pendingAck` set. New messages from consumer are checked against this set and considered duplicate.

Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: It was just an effort made by JMS Input to delete the last saved artifact because it ack failed. It didn't rely on that call to be successful. If we add an end marker for window then we need to add an additional method to WindowDataManager api and every operator needs to invoke it after saving a window.

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 9 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java:409]: JMS Input in its endWindow does the following:
1. saves the artifact
2. acks the session

When acks fails but the artifact got saved, then it tries to delete the artifact.
If the artifact is deleted successfully, the window is considered completed or not.
1. When deleted - window was not completed and the messages will be consumed from JMS
2. When not deleted - window was considered completed so it is replayed from file. However while replaying message ids are added to pendingAck set. When these messages are received from JMS again (since ack failed earlier), they will be considered re-delivered.

Deletion of a window during `setup` is difficult because we do not actually know if ack failed for that window.

**Comment by tweise** [comment]
Build fails due to addition from another PR. Please rebase and squash.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java [hunk 15]
```diff
@@ -177,25 +179,50 @@ public void save(Map<Long, Map<Slice, Bucket.BucketedValue>> unsavedData, int op
...
+    Map<Long, Object> artifactPerWindow = new HashMap<>();
+    FileSystemWAL.FileSystemWALReader reader = getWal().getReader();
+    reader.seek(getWal().getWalStartPointer());
+    
+    Slice windowSlice = readNext(reader);
+    while (reader.getCurrentPointer().compareTo(getWal().getWalEndPointerAfterRecovery()) < 0 && windowSlice != null) {
+      long window = Longs.fromByteArray(windowSlice.toByteArray());
+      Object data = fromSlice(readNext(reader));
+      windowSlice = readNext(reader); //null, deleted or next window
+      if (windowSlice == DELETED) {
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java]: Won't this always be 2 different objects?

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 10 above)
Comment (tweise) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: So does this mean the delete isn't really needed? What is the benefit of keeping it? It would be good to document it in the JMS operator. Also it sounds like the DELETED logic is very specific to the JMS operator. If it is not needed anywhere else, then how about pulling it into the JMS operator?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java (hunk 15 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java]: Yeah wonder why the test didn't catch this. Will check

File: library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java (hunk 10 above)
Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Will make this change

Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Is it okay if this delete logic is removed from JMS as well? 
It is difficult to pull it into JMS because the api for saving is `save(window id,Object artifact)`. 
If JMSInput calls `save(windowId, DELETED)` then that creates multiple entries for the same window id. So in order for JMSInput to figure out if the last window is deleted or not, it has to read that window twice  which I think is hacky.

Either we completely remove this functionality  or we change WindowDataManager Api to have a trigger for `beginWindow(long window)` in which it can record the window id to wal and `save` just saves the artifact not the window.

Comment (tweise) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: What's the downside of removing it completely?

Comment (otterc) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Don't think there is any. If the ack failed but the state is saved, the messages are re-played from the state. However since the ack failed, they will also be re-delivered by the message bus but we ignore them. So we do not send any duplicates downstream or miss any message .

Comment (tweise) [reply library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java]: Saving the state and detecting the duplicate is how I would expect it to work. So +1 on removing this logic altogether. It would be good to also document this in the JMSInputOperator.

**Comment by otterc** [comment]
Addressed all the comments
